### PR TITLE
Coroutine debugger optimization

### PR DESCRIPTION
**Pull request by nikita-nazarov**
Add a method that allows dumping all coroutine information needed by the IDEA debugger in one call.

### CODE REVIEW

File: kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt [hunk 1]
```diff
@@ -163,6 +163,50 @@ internal object DebugProbesImpl {
...
+        val lastObservedFrameRefs = mutableListOf<CoroutineStackFrame?>()
+        val coroutinesInfoAsJson = mutableListOf<String>()
+        val coroutinesInfo = dumpCoroutinesInfo()
+        for (info in coroutinesInfo) {
+            val context = info.context
+            coroutinesInfoAsJson.add(
+                """
+                {
+                   name: ${context[CoroutineName.Key]?.name?.surroundWithQuotes()},
+                   id: ${context[CoroutineId.Key]?.id},
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt:191]: This is not a real JSON: keys are not quoted. Either it should be documented (e.g. an additional comment) or fixed

File: kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt [hunk 2]
```diff
@@ -163,6 +163,50 @@ internal object DebugProbesImpl {
...
+     */
+    @OptIn(ExperimentalStdlibApi::class)
+    public fun dumpCoroutinesInfoAsJSONAndReferences(): Array<Any> {
+        fun String.surroundWithQuotes() =
+            "\"$this\""
+
+        val lastObservedThreadRefs = mutableListOf<Thread?>()
+        val lastObservedFrameRefs = mutableListOf<CoroutineStackFrame?>()
+        val coroutinesInfoAsJson = mutableListOf<String>()
+        val coroutinesInfo = dumpCoroutinesInfo()
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt:184]: [nit] you can first dump coroutines and then pre-allocate three lists with the known size

File: kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt [hunk 3]
```diff
@@ -163,6 +163,50 @@ internal object DebugProbesImpl {
...
+     * to a client VM to speed up fetching of coroutine information.
+     * The return value is an array of objects, which consists of four elements:
+     * 1) A string in a JSON format that stores information that is needed to display
+     *    every coroutine in the coroutine panel in the IDEA debugger.
+     * 2) An array of last observed threads.
+     * 3) An array of last observed frames.
+     * 4) An array of DebugCoroutineInfo.
+     */
+    @OptIn(ExperimentalStdlibApi::class)
+    public fun dumpCoroutinesInfoAsJSONAndReferences(): Array<Any> {
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt:177]:
```suggestion
    public fun dumpCoroutinesInfoAsJsonAndReferences(): Array<Any> {
```

Comment (qwwdfsad) [reply kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt:177]: Please add a few unit-tests that ensure the resulting format, for empty list of coroutines and for some coroutines (mixing `null` and non-null values)

File: kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt [hunk 4]
```diff
@@ -163,6 +163,50 @@ internal object DebugProbesImpl {
...
+    /*
+     * This method optimises the number of packages sent by the IDEA debugger
+     * to a client VM to speed up fetching of coroutine information.
+     * The return value is an array of objects, which consists of four elements:
+     * 1) A string in a JSON format that stores information that is needed to display
+     *    every coroutine in the coroutine panel in the IDEA debugger.
+     * 2) An array of last observed threads.
+     * 3) An array of last observed frames.
+     * 4) An array of DebugCoroutineInfo.
+     */
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/jvm/src/debug/internal/DebugProbesImpl.kt:175]:
```suggestion
     *
     * Internal (JVM-public) method used by IDEA debugger as of 1.6.0-RC. 
     */
```

**Review by qwwdfsad — CHANGES_REQUESTED** [review]

**Comment by qwwdfsad** [comment]
Thanks!

Will merge manually here: https://github.com/Kotlin/kotlinx.coroutines/pull/2923
